pkg/controller/pdb.go

```diff
@@ -0,0 +1,287 @@
+// as published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License, version 3,
+// along with this program.  If not, see <http://www.gnu.org/licenses/>
+
+package controller
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"sync"
+
+	v2 "github.com/minio/operator/pkg/apis/minio.min.io/v2"
+	v1 "k8s.io/api/policy/v1"
+	"k8s.io/api/policy/v1beta1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/klog/v2"
+)
+
+// DeletePDB - delete PDB for tenant
+func (c *Controller) DeletePDB(ctx context.Context, t *v2.Tenant) (err error) {
+	available := c.GetPDBAvailable()
+	if !available.Available() {
+		return nil
+	}
+	if available.V1Available() {
+		err := c.kubeClientSet.PolicyV1().PodDisruptionBudgets(t.Namespace).DeleteCollection(ctx, metav1.DeleteOptions{}, metav1.ListOptions{
+			LabelSelector: metav1.SetAsLabelSelector(labels.Set{
+				v2.TenantLabel: t.Name,
+			}).String(),
+		})
+		if err != nil {
+			// don't exist
+			if k8serrors.IsNotFound(err) {
+				return nil
+			}
+			klog.Errorf("Delete tenant %s's V1.PDB failed:%s", t.Name, err.Error())
+			return err
+		}
+	}
+	if available.V1BetaAvailable() {
+		err := c.kubeClientSet.PolicyV1beta1().PodDisruptionBudgets(t.Namespace).DeleteCollection(ctx, metav1.DeleteOptions{}, metav1.ListOptions{
+			LabelSelector: metav1.SetAsLabelSelector(labels.Set{
+				v2.TenantLabel: t.Name,
+			}).String(),
+		})
+		if err != nil {
+			// don't exist
+			if k8serrors.IsNotFound(err) {
+				return nil
+			}
+			klog.Errorf("Delete tenant %s's V1Beta.PDB failed:%s", t.Name, err.Error())
+			return err
+		}
+	}
+	return nil
+}
+
+// CreateOrUpdatePDB - hold PDB as expected
+func (c *Controller) CreateOrUpdatePDB(ctx context.Context, t *v2.Tenant) (err error) {
+	available := c.GetPDBAvailable()
+	if !available.Available() {
+		return nil
+	}
+	for _, pool := range t.Spec.Pools {
+		if strings.TrimSpace(pool.Name) == "" {
+			continue
+		}
+		if available.Available() {
+			// check sts status first.
+			ssName := t.PoolStatefulsetName(&pool)
+			existingStatefulSet, err := c.statefulSetLister.StatefulSets(t.Namespace).Get(ssName)
+			if err != nil {
+				return err
+			}
+			if existingStatefulSet.Status.ReadyReplicas != existingStatefulSet.Status.Replicas || existingStatefulSet.Status.Replicas == 0 {
+				return nil
+			}
+			if t.Status.CurrentState != StatusInitialized {
+				return nil
+			}
+		}
+		if available.V1Available() {
```

Comment:
Logic is same for both cases (other than pdb object). Can we create a function and use for both scenarios?

Reply:
I've tried it and can't implement it, the Client of these two is different, and the similarity of the incoming parameter is also different. @shtripat If you have a good thought you can share it with me as well

Reply:
yes, after closure look, it doesn't look feasible. anyway its not a major one.